internals/overlord/state/notices.go

```diff
@@ -0,0 +1,386 @@
+	}
+	n.id = jn.ID
+	n.noticeType = NoticeType(jn.Type)
+	n.key = jn.Key
+	n.firstOccurred = jn.FirstOccurred
+	n.lastOccurred = jn.LastOccurred
+	n.lastRepeated = jn.LastRepeated
+	n.occurrences = jn.Occurrences
+	n.lastData = jn.LastData
+	if jn.RepeatAfter != "" {
+		n.repeatAfter, err = time.ParseDuration(jn.RepeatAfter)
+		if err != nil {
+			return fmt.Errorf("invalid repeat-after duration: %w", err)
+		}
+	}
+	if jn.ExpireAfter != "" {
+		n.expireAfter, err = time.ParseDuration(jn.ExpireAfter)
+		if err != nil {
+			return fmt.Errorf("invalid expire-after duration: %w", err)
+		}
+	}
+	return nil
+}
+
+type NoticeType string
+
+const (
+	// Recorded whenever a change is updated: when it is first spawned or its
+	// status was updated.
+	NoticeChangeUpdate NoticeType = "change-update"
+
+	// A client notice reported via the Pebble client API or "pebble notify".
+	// The key and data fields are provided by the user. The key must be in
+	// the format "mydomain.io/mykey" to ensure well-namespaced notice keys.
+	NoticeClient NoticeType = "client"
+
+	// Warnings are a subset of notices where the key is a human-readable
+	// warning message.
+	NoticeWarning NoticeType = "warning"
+)
+
+// NoticeTypeFromString validates the given string and returns the NoticeType,
+// or empty string if it's not valid.
+func NoticeTypeFromString(s string) NoticeType {
+	noticeType := NoticeType(s)
+	switch noticeType {
+	case NoticeChangeUpdate, NoticeClient, NoticeWarning:
+		return noticeType
+	default:
+		return ""
+	}
+}
+
+// AddNotice adds an occurrence of a notice with the specified type and key
+// and key-value data, returning the notice ID.
+func (s *State) AddNotice(noticeType NoticeType, key string, data map[string]string, repeatAfter time.Duration) string {
+	return s.addNoticeWithTime(time.Now(), noticeType, key, data, repeatAfter)
+}
+
+func (s *State) addNoticeWithTime(now time.Time, noticeType NoticeType, key string, data map[string]string, repeatAfter time.Duration) string {
+	if noticeType == "" || key == "" || len(key) > MaxNoticeKeyLength {
+		// Programming error (max key length has already been checked by API)
+		logger.Panicf("Internal error, please report: attempted to add invalid notice (type %q, key %q)",
+			noticeType, key)
+	}
+
+	s.writing()
+
+	now = now.UTC()
+	newOrRepeated := false
+	uniqueKey := uniqueNoticeKey(noticeType, key)
+	notice, ok := s.notices[uniqueKey]
+	if !ok {
+		// First occurrence of this notice type+key
+		s.lastNoticeId++
+		notice = &Notice{
+			id:            strconv.Itoa(s.lastNoticeId),
+			noticeType:    noticeType,
+			key:           key,
+			firstOccurred: now,
+			lastRepeated:  now,
+			repeatAfter:   repeatAfter,
+			expireAfter:   noticeExpireAfter,
+			occurrences:   1,
+		}
+		s.notices[uniqueKey] = notice
+		newOrRepeated = true
+	} else {
+		// Additional occurrence, update existing notice
+		notice.occurrences++
+		if repeatAfter != 0 && now.After(notice.lastRepeated.Add(repeatAfter)) {
```

Comment:
Good call. I've tweaked how this works, so that the `repeatAfter` time is set/updated only if it's nonzero. So you can call `AddNotice()` with a specific repeatAfter time, and then call it with repeatAfter=0 on subsequent calls if you want. We probably won't in Pebble code itself (we'll probably use the same repeatAfter value each call), but it seems useful behaviour for client notices added using `pebble notify`.